Expose LayerFactory::LayerTypeList in pycaffe

lukeyeager opened the pull request:
I want to be able to validate NetParameters with pycaffe without waiting for Caffe to crash on SIGABRT.
```
F0814 16:58:06.833762 31396 layer_factory.hpp:80] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: not-a-layer-type (known types: AbsVal, Accuracy, ArgMax, BNLL, Concat, ContrastiveLoss, Convolution, Data, Deconvolution, Dropout, DummyData, Eltwise, EuclideanLoss, Exp, Filter, Flatten, HDF5Data, HDF5Output, HingeLoss, Im2col, ImageData, InfogainLoss, InnerProduct, LRN, Log, MVN, MemoryData, MultinomialLogisticLoss, PReLU, Pooling, Power, Python, ReLU, Reduction, Reshape, SPP, Sigmoid, SigmoidCrossEntropyLoss, Silence, Slice, Softmax, SoftmaxWithLoss, Split, TanH, Threshold, WindowData)
*** Check failure stack trace: ***
Aborted (core dumped)
```
Exposing `LayerTypeList` in pycaffe lets me check the layer types and handle errors in a reasonable way.

Is there any way to catch all CHECK/DCHECK errors with pycaffe and turn them into Python exceptions?

ronghanghu commented:
Look good to me :)

jeffdonahue commented on python/caffe/test/test_layer_type_list.py:
```diff
@@ -0,0 +1,10 @@
+
+    def test_standard_types(self):
+        for type_name in ['Data', 'Convolution', 'InnerProduct']:
+            assert type_name in caffe.layer_type_list(), \
```
this should use `self.assertTrue(...)` instead of `assert`

jeffdonahue commented:
LGTM besides above comment.

lukeyeager commented:
Thanks for the review. Changed and squashed. Why is `unittest.assertIn()` preferable to `assert`?

Also, can you address this question:

> Is there any way to catch all CHECK/DCHECK errors with pycaffe and turn them into Python exceptions?

That would be a huge help.

jeffdonahue commented:
Thanks for the update! The `unittest.assert*` functions will make the test pass/fail appropriately according to their result and print a nicely formatted error message when the assertion fails, where `assert` will just raise an Exception if its argument is false. (Maybe the exception would also make the test fail, but presumably with a less nice error message.)

> Is there any way to catch all CHECK/DCHECK errors with pycaffe and turn them into Python exceptions?

I think #1683 tried to do this
